### PR TITLE
internal/v4: implement archive PUT

### PR DESCRIPTION
**Pull request by rogpeppe**
This will allow the ingester to match the uploaded charm URLs to the original
launchpad (and therefore legacy charm store) revisions.

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/301/

**Comment by mhilton** [comment]
:+1:  LGTM

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/303/

File: internal/v4/archive.go [hunk 1]
```diff
@@ -79,19 +84,20 @@ func (h *Handler) serveDeleteArchive(id *charm.Reference, w http.ResponseWriter,
 	return nil
 }
 
+func (h *Handler) updateStatsArchiveUpload(id *charm.Reference, err *error) {
+	// Upload stats don't include revision: it is assumed that each
+	// entity revision is only uploaded once.
+	id.Revision = -1
```
Comment (urosj) [inline internal/v4/archive.go:96]: Good catch!

File: internal/v4/archive.go [hunk 2]
```diff
@@ -113,15 +119,58 @@ func (h *Handler) servePostArchive(id *charm.Reference, w http.ResponseWriter, r
...
+	}
+	if err := h.addBlobAndEntity(id, req.Body, hash, req.ContentLength); err != nil {
+		return errgo.Mask(err, errgo.Is(params.ErrDuplicateUpload))
+	}
+	return router.WriteJSON(w, http.StatusOK, &params.ArchiveUploadResponse{
+		Id: id,
+	})
+}
+
+func (h *Handler) servePutArchive(id *charm.Reference, w http.ResponseWriter, req *http.Request) (err error) {
```
Comment (urosj) [inline internal/v4/archive.go:147]: Please add comment what the purpose and usage of PUT is.

**Comment by urosj** [comment]
:+1: Thank you.

File: internal/v4/archive.go (hunk 2 above)
Comment (rogpeppe) [inline internal/v4/archive.go:147]: done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/305/

**Pull request merged**